test_programs/execution_success/array_regex/src/main.nr

```diff
@@ -0,0 +1,505 @@
+        }
+    }
+}
+
+struct Question<T> {
+    inner: T,
+}
+
+impl<T> Regex for Question<T> where T: Regex {
+    fn match<let N: u32>(self, input: Bvec<u8, N>) -> Match<N> {
+        Or {
+            lhs: self.inner,
+            rhs: (),
+        }.match(input)
+    }
+}
+
+// 0 <= num_matches <= N
+struct Star<T, let N: u32> {
+    inner: T,
+}
+
+impl<T, let M: u32> Regex for Star<T, M> where T: Regex {
+    fn match<let N: u32>(self, input: Bvec<u8, N>) -> Match<N> {
+        let regex: Repeated<_, M> = Repeated {
+            inner: Question { inner: self.inner },
+        };
+        regex.match(input)
+    }
+}
+
+// 0 < num_matches <= N
+struct Plus<T, let N: u32, let N_PRED: u32> {
+    inner: T,
+}
+
+impl<T, let M: u32, let M_PRED: u32> Regex for Plus<T, M, M_PRED> where T: Regex {
+    fn match<let N: u32>(self, input: Bvec<u8, N>) -> Match<N> {
+        std::static_assert(M_PRED + 1 == M, "M - 1 != M_PRED");
+        let star: Star<T, M_PRED> = Star { inner: self.inner };
+        (
+            self.inner,
+            star
+        ).match(input)
+    }
+}
+
+// Repeated is to (,) as AnyOf is to Or
+struct AnyOf<T, let N: u32> {
+    inner: [T; N],
+}
+
+impl<T, let M: u32> Regex for AnyOf<T, M> where T: Regex {
+    fn match<let N: u32>(self, input: Bvec<u8, N>) -> Match<N> {
+        let mut result = Match::failed(input);
+        for i in 0..M {
+            if !result.succeeded {
+                result = self.inner[i].match(result.leftover);
+            }
+        }
+        result
+    }
+}
+
+fn reverse_array<T, let N: u32>(input: [T; N]) -> [T; N] {
+    let mut output = [std::unsafe::zeroed(); N];
+    for i in 0..N {
+        output[i] = input[N - (i + 1)];
+    }
+    output
+}
+
+fn main() {
+    assert_eq(reverse_array([1, 2, 3, 4]), [4, 3, 2, 1]);
+
+    let mut xs: Bvec<u8, 3> = Bvec::empty();
+
+    xs = xs.push(0);
+    assert_eq(xs, Bvec { inner: [0, 0, 0], offset: 0, len: 1 });
+
+    xs = xs.push(1);
+    assert_eq(xs, Bvec { inner: [0, 1, 0], offset: 0, len: 2 });
+
+    xs = xs.push(2);
+    assert_eq(xs, Bvec { inner: [0, 1, 2], offset: 0, len: 3 });
+
+    xs = xs.push(3);
+    assert_eq(xs, Bvec { inner: [0, 1, 2], offset: 0, len: 3 });
+
+    let ys = Bvec::new([0, 1, 2]);
+    assert_eq(xs, ys);
+
+    // test that pop_front gives all contents, in order,
+    // followed by std::unsafe::zeroed()
+    println(xs);
+    let (x, new_xs) = xs.pop_front();
+    assert_eq(x, 0);
+
+    xs = new_xs;
+    println(xs);
+    let (x, new_xs) = xs.pop_front();
+    assert_eq(x, 1);
+
+    xs = new_xs;
+    println(xs);
+    let (x, new_xs) = xs.pop_front();
+    assert_eq(x, 2);
+
+    xs = new_xs;
+    println(xs);
+    if xs.len != 0 {
+        let (x, _new_xs) = xs.pop_front();
+        assert_eq(x, std::unsafe::zeroed());
+    }
+
+    assert_eq(new_xs, Bvec { inner: [0, 1, 2], offset: 3, len: 0 });
+
+    // gr(a|e)y
+    let graey_regex = ("gr", (Or { lhs: "a", rhs: "e" }, "y"));
```

Comment:
I wonder if we could make the API even nicer with something like:
```
impl<T> for T where T: Regex {
    fn or<U>(self, other: U) -> (Self, U) {
        Or { lhs: self, rhs: other }
    }
}

// ("gr", "a".or("e"), "y")
```
Not sure that we support this though.

Reply:
R.e. the `Or`: that would be cool, but it looks like we don't support it. I tried a couple versions and ran into `error: Non-struct type used in impl`.

Perhaps a version defined over `impl Regex` instead of `T` would be possible, but we have limited support for `TraitAsType`.

R.e. `(A, B, C)` as `((A, B), C)`: that's a convenient feature that I left out to keep the test shorter. I'd add it if pulling this out into a separate repo.